src/Assistant.ts

```diff
@@ -0,0 +1,387 @@
+import {
+  AssistantThreadsSetStatusResponse,
+  AssistantThreadsSetSuggestedPromptsResponse,
+  AssistantThreadsSetTitleResponse,
+  ChatPostMessageArguments,
+} from '@slack/web-api';
+import processMiddleware from './middleware/process';
+import {
+  AllMiddlewareArgs,
+  AnyMiddlewareArgs,
+  Context,
+  Middleware,
+  SayFn,
+  SlackEventMiddlewareArgs,
+} from './types';
+import { AssistantInitializationError, AssistantMissingPropertyError } from './errors';
+import {
+  AssistantThreadContext,
+  AssistantThreadContextStore,
+  DefaultThreadContextStore,
+  GetThreadContextFn,
+  SaveThreadContextFn,
+} from './AssistantThreadContextStore';
+
+/**
+ * Configuration object used to instantiate the Assistant
+ */
+export interface AssistantConfig {
+  threadContextStore?: AssistantThreadContextStore;
+  threadStarted: AssistantThreadStartedMiddleware | AssistantThreadStartedMiddleware[];
+  threadContextChanged?: AssistantThreadContextChangedMiddleware | AssistantThreadContextChangedMiddleware[];
+  userMessage: AssistantUserMessageMiddleware | AssistantUserMessageMiddleware[];
+}
```

Comment:
This pull request does not support Block Kit interactions + botMessage listener. For initial release, that's fine, but we may receive feature requests to add the advanced feature, which is available in bolt-python and bolt-java

Reply:
Can you point me to what you're referring to, specifically within the Python implementation? Without looking at the code, I presume you mean when a user interacts with a BKE, that is also directed to this listener (rather than the other middleware that exists for it)?

Reply:
I had a look, but all I see in that example appears to be a hand-rolled approach that combines the manual adding, parsing, and handling of metadata paired with the use of regular listeners.

What about the current implementation would prevent that same approach from being used? What is not being handled by Bolt behind the scenes and on behalf of the developer that should?

Reply:
Your current implementation does not provide a way to handle the app's bot message events. Something equivalent to ignoring_self_assistant_message_events_enabled=False is necessary if a developer would like to post a message with user inputs first, and then continue its processing in a following listener (botMessage listener). It's still feasible to do eveything within app.view listener but the user experience is not optimal plus the code can be less simple.

Reply:
I've thought about this a bit longer, and if it's OK with you, I'd like to avoid introducing more conveniences until we know how folks are actually using this. If @filmaj finishes his review and disagrees, I'll introduce it, but I'm concerned that we're already making this quite complicated/opinionated out of the box before anyone has used it or requested additional features.

Reply:
A fast follow seems reasonable to me 🤷 ; also I have a large PR (bolt v4 https://github.com/slackapi/bolt-js/pull/2254) that is waiting on this to be merged, so I selfishly would like to prevent adding more until I can reconcile these two (which will require a bit of work AFAICT)

Reply:
This is definitely an advanced usage of the feature, so I don't think it's a release blocker for the initial rollout. Also, nothing prevents us from adding the feature later on. We can add it in future releases without any issues.